lib/storage.js

```diff
@@ -13,47 +13,98 @@
 const http = require('./storage/http');
 const datauri = require('./storage/datauri');
 const URL = require('url');
 const fs = require("fs-extra");
 const path = require('path');
+const mime = require('mime-types');
+const validUrl = require('valid-url');
+const url = require('url');
 
-const DATA_PROTOCOL = 'data:';
+const SOURCE_BASENAME = 'source';
 
-async function getSource(paramsSource, inDirectory, disableSourceDownload) {
-    // Note: validation has happened in validate.js before
+const DATA_PROTOCOL = 'data:';
 
+async function getAsset(assetReference, directory, name, disableDownload) {
+    // normalize asset reference to be an object
+    if(!assetReference) {
+        throw new GenericError('Missing assetReference');
+    }
+    if (typeof assetReference === 'string') {
+        assetReference = { url: assetReference };
+    }
     if (process.env.WORKER_TEST_MODE) {
         // local file support for `Asset Compute test-worker` unit tests
         // not supported for clients in production
         // params.source.url will just be a filename like 'file.jpg'
-        const filePath = path.join(inDirectory, paramsSource.url);
+        name = assetReference.url;
+        const filePath = path.join(directory, name);
         if (!fs.existsSync(filePath)) {
-            throw new Error(`Invalid or missing local file ${paramsSource.url}`);
+            throw new Error(`Invalid or missing local file ${name}`);
         }
-        return new Source(paramsSource.url, inDirectory);
+        return new Source(assetReference, directory, name);
 
     } else {
-        const source = new Source(paramsSource, inDirectory);
-        if (disableSourceDownload) {
+        const source = new Source(assetReference, directory, name);
+        if (disableDownload) {
             console.log(`Skipping source file download for ${AssetComputeLogUtils.redactUrl(source.url)}`);
         } else {
             // TODO: error handling: errors catched & thrown in http.download() and then catched again in worker.js ???
             // download http/https url into file
-            const protocol = URL.parse(paramsSource.url).protocol;
+            const protocol = URL.parse(assetReference.url).protocol;
             if (protocol === DATA_PROTOCOL) {
-                await datauri.download(paramsSource, source.path);
+                await datauri.download(assetReference, source.path);
             } else {
-                await http.download(paramsSource, source.path);
+                await http.download(assetReference, source.path);
             }
         }
 
         return source;
     }
 }
 
+// function to return an extension for a file
+// if not empty returns a leading period
+// prefers extension from the file over name determined by mimeType
+function extension(filename, mimeType) {
+    let ext = '';
+    if (filename) {
+        ext = path.extname(filename);
+    }
+    if (!ext && mimeType) {
```

Comment:
@alexkli I wonder if we should test mimetype first to rather rely on mimetype than extension... ? (Possibly later).

Reply:
This is just for the "nice" source extension used in `sourceFilename()`... see also comment below.